pkg/morph/deploy/notary.go

```diff
@@ -869,3 +878,353 @@ func makeUnsignedDesignateCommitteeNotaryTx(roleContract *rolemgmt.Contract, com
+							zap.Uint32("current height", currentHeight), zap.Uint32("retry after height", expirationTxValidUntilBlock))
+						return
+					}
+
+					l.Info("previously sent transaction increasing expiration time of the Notary deposit expired without side-effect ")
+				}
+
+				l.Info("sending new transaction increasing expiration time of the Notary deposit...", zap.Uint32("till", till))
+
+				_, vub, err := notaryContract.LockDepositUntil(localAccID, till)
+				if err != nil {
+					l.Error("failed to send transaction increasing expiration time of the Notary deposit, will try again later", zap.Error(err))
+					return
+				}
+
+				expirationTxValidUntilBlock = vub
+
+				l.Info("transaction increasing expiration time of the Notary deposit has been successfully sent, will wait for the outcome")
+
+				return
+			}
+		}
+
+		if transferTxValidUntilBlock > 0 {
+			l.Info("transaction transferring local account's GAS to the Notary contract was sent earlier, checking relevance...")
+
+			// for simplicity, we track ValidUntilBlock. In this particular case, it'd be
+			// more efficient to monitor a transaction by ID, because side effect is
+			// inconsistent (funds can be spent in background).
+
+			if currentHeight <= transferTxValidUntilBlock {
+				l.Info("previously sent transaction transferring local account's GAS to the Notary contract may still be relevant, will wait for the outcome",
+					zap.Uint32("current height", currentHeight), zap.Uint32("retry after height", transferTxValidUntilBlock))
+				return
+			}
+
+			l.Info("previously sent transaction transferring local account's GAS to the Notary contract expired without side-effect")
+		}
+
+		needAtLeast := new(big.Int).Mul(singleNotaryDepositAmount, big.NewInt(50))
+		if gasBalance.Cmp(needAtLeast) < 0 {
+			l.Info("minimum threshold for GAS transfer from local account to the Notary contract not reached, will wait for replenishment",
+				zap.Stringer("need at least", needAtLeast), zap.Stringer("have", gasBalance))
+			return
+		}
+
+		var transferData notary.OnNEP17PaymentData
+		transferData.Account = &localAccID
+		transferData.Till = till
+
+		l.Info("sending new transaction transferring local account's GAS to the Notary contract...",
+			zap.Stringer("amount", singleNotaryDepositAmount), zap.Uint32("till", transferData.Till))
+
+		// nep17.TokenWriter.Transfer doesn't support notary.OnNEP17PaymentData
+		// directly, so split the args
+		// Track release of https://github.com/nspcc-dev/neo-go/issues/2987
+		_, vub, err := gasContract.Transfer(localAccID, notary.Hash, singleNotaryDepositAmount, []interface{}{transferData.Account, transferData.Till})
+		if err != nil {
+			l.Error("failed to send transaction transferring local account's GAS to the Notary contract, will try again later", zap.Error(err))
+			return
+		}
+
+		transferTxValidUntilBlock = vub
```

Comment:
Just wait for it and pass the context if you need an early stop. You will update contracts one-by-one, thus no concurrency issues are expected.

Reply:
same https://github.com/nspcc-dev/neofs-node/pull/2408#discussion_r1250888297

Reply:
yeah, i've written a comment about separate routine which controls only by itself. Do u agree it's worth to implement it? And if so, how do u see its behavior? My thoughts were:
1. just async track Notary balance (e.g. on new block receival) and make sure it doesn't go below the threshold (e.g. 1GAS)
2. wait for external signals from other routines (any notary action during procedure) and refresh deposit only by signal
3. maybe some other alternative?

@AnnaShaleva im not strong enought in Notary system, so full trust to u :grinning:

Reply:
Discussed, neet to try out waiter in a separate commit and create an issue afterwards.

Reply:
added commit, check it out pls